### DISCUSSION TITLE
iPXE boot of KVM guest fails: Initramfs unpacking failed: broken padding

### DISCUSSION THREAD

**Discussion by JKDingwall** [ipxe/ipxe#1244] [General]
Hi,

We have had intermittent issues in the past trying to boot under KVM (iPXE chainload) vs physical where the KVM boot would fail but the physical would succeed.  This could usually be resolved by rebuilding the initramfs image such that it ended up with a different size.  Today we've got a new problem with an environment using the Ubuntu `Linux kernel x86 boot executable bzImage, version 5.15.0-113-generic (buildd@lcy02-amd64-017) #123~20.04.1-Ubuntu SMP Wed Jun 12 17:33:13 UTC 2024` kernel where it fails with the message in the discussion title.  I have two workarounds:

1. reorder artefact download from kernel / initramfs (main) / initramfs (supplemetary cpio) to initramfs (main) / initramfs (supplemetary cpio) / kernel
2. pad the kernel to a 4k boundary with `truncate -s 11550720 vmlinuz` and keep the original order (original kernel size 11548296)

My suspicion is that somehow the combination of the in memory layout of the kernel / initramfs files produced by iPXE and the kernel decoding routine results in the initramfs being missed.  Are there any further suggestions on the debugging which can be carried out from here?

Two iPXE versions were tested: tag v1.21.1 and commit e965f179e1654103eca33feed7a9cc4c51d91be6

Full iPXE script:

```
#!ipxe
echo iPXE Boot String
show unixtime
kernel https://<boot_server_ip>:8080/vmlinuz initrd=certs.cpio initrd=initrd.img BOOTIF=<client_mac> ip=<client_ip>::<client_gw>:<netmask>:<hostname>::off netboot=http boot=casper locale=en_GB.UTF-8 setkmap=gb mirror/country=GB hostname=<hostname> fetch=https://<boot_server_ip>:8080/filesystem.squashfs audit=1 quiet splash  
imgverify vmlinuz https://<boot_server_ip>:8080/vmlinuz.sig

initrd https://<boot_server_ip>:8080/initrd.img
imgverify initrd.img https://<boot_server_ip>:8080/initrd.img.sig

initrd https://<boot_server_ip>:8080/<client_mac>/certs.cpio
imgverify certs.cpio https://<boot_server_ip>:8080/<client_mac>/certs.cpio.sig
boot ||
# If everything failed, give the user some options
echo Boot failed
prompt --key 0x02 --timeout 2000 Press Ctrl-B for the iPXE command line... && shell ||
```

**Comment by NiKiZe** [comment]
As promised in https://github.com/ipxe/ipxe/issues/1312#issuecomment-2382603700

There is a few things to check.
* What does `imgstat` show before `boot` ?
* exactly which kernels are you seeing issues with? (Use `uname -a` with the booted kernel)
* Are you booting `pcbios` or `efi` ?

For `pcbios` the initrd images are concatenated, and the initrd=xxx of the kernel command line has no effect.

For efi and newer kernel there is some magic doing the same, but initrd= might have an effect. Older kernels in EFI they are required.

**Reply by JKDingwall** [reply]
I will have to set the environment up again to get all of the answers.  The affected kernel version was Ubuntu's 5.15.0-113-generic for 20.04/focal.  We had similar previous issues where changing the size of the main initrd resolved the boot error however in this case it seemed it was the kernel that was significant.  The systems are booting using `pcbios`:

```
# file undionly.kpxe 
undionly.kpxe: pxelinux loader (version 3.70 or newer)
```

I don't believe it to be a kernel bug since the behaviour with the same kernel/initrd was different between physical and virtual.  I suppose that the virtualisation beneath iPXE could be the culprit though.

**Reply by NiKiZe** [reply]
So pcbios, as I mentioned, that should mean that your initrd files, from the point of the kernel, is just concatenated together.

`imgstat` can verify which files that is, so you don't have a script or background image or similar.

Other than that possibly memory layout, but highly doubt that.
